thalo-eventstoredb/src/event_store.rs

```diff
@@ -0,0 +1,340 @@
+            .position(StreamPosition::End)
+            .max_count(1);
+
+        let events = self.read_stream(self.stream_id::<A>(Some(id)), options).await?;
+        if let Some(event) = events.iter().next() {
+            let event_data = event.get_original_event();
+            return Ok(Some(event_data.revision as usize));
+        }
+
+        Ok(None)
+    }
+
+    async fn save_events<A>(
+        &self,
+        id: &<A as Aggregate>::ID,
+        events: &[<A as Aggregate>::Event],
+    ) -> Result<Vec<usize>, Self::Error>
+    where
+        A: Aggregate,
+        <A as Aggregate>::Event: serde::Serialize,
+    {
+        if events.is_empty() {
+            return Ok(vec![]);
+        }
+
+        let sequence = self.load_aggregate_sequence::<A>(id).await.unwrap_or(None);
+
+        let revision = {
+            match sequence {
+                Some(n) => ExpectedRevision::Exact(n as u64),
+                None => ExpectedRevision::NoStream,
+            }
+        };
+
+        let mut payload: Vec<EventData> = vec![];
+
+        for event in events.iter() {
+            let event_data_payload = ESDBEventPayload {
+                created_at: Utc::now(),
+                aggregate_type: <A as TypeId>::type_id().to_string(),
+                aggregate_id: id.to_string(),
+                event_data: serde_json::to_value(event).map_err(Error::SerializeEvent)?,
+            };
+
+            let event_data = EventData::json(event.event_type(), &event_data_payload)
+                .map_err(Error::SerializeEventDataPayload)?
+                .id(Uuid::new_v4());
+
+            payload.push(event_data);
+        }
+
+        let options = AppendToStreamOptions::default().expected_revision(revision);
+        let res = self
+            .client
+            .append_to_stream(&self.stream_id::<A>(Some(id)), &options, payload)
+            .map_err(|err| Error::WriteStreamError(sequence.unwrap_or(0) as usize, err))
+            .await?;
+
+        let end_position = res.next_expected_version as usize;
+        let start_position = end_position - (events.len() - 1);
+
+        Ok((start_position..end_position + 1).collect::<Vec<usize>>())
+    }
+}
+
+impl Debug for ESDBEventStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ESDBEventStore")
+            .field("client", &"eventstore::Client")
+            .finish()
+    }
+}
+
+#[cfg(feature = "debug")]
+impl ESDBEventStore {
+    pub async fn print<A>(&self)
+    where
+        A: Aggregate,
+        <A as Aggregate>::Event: serde::Serialize,
+    {
+        let stream_res = self.client.read_all(&Default::default()).await;
+
+        let mut events: Vec<(Uuid, u64, ESDBEventPayload)> = vec![];
+        if let Ok(mut stream) = stream_res {
+            while let Some(event) = stream.next().await.unwrap() {
+                let event_data = event.get_original_event();
+
+                if event_data.event_type.starts_with("$") {
```

Comment:
You might consider using server-side filtering so you don't waste bandwidth for data you don't want: https://developers.eventstore.com/clients/grpc/subscriptions.html#server-side-filtering

Reply:
Squash that, you are only interested in doing simple reads. I missed that part.

Reply:
Thanks! That feature will actually come in handy with some other features I'm working on in the future and I wasn't aware of it before.